Transformer Modules

Atticus1806 opened the pull request:
Fix #53.

This is a draft for now, since Attention Modules from #52 need to be implemented for that. Comments on code style, changes etc. are welcome already of course. 

Orienting on PyTorch for naming of variables and functions.
[PyTorch documentation](https://pytorch.org/docs/stable/nn.html#transformer-layers).

Note that PyTorch also has a masking logic in almost all modules, which was left out for now, since it is used in the Attention Modules in PyTorch. Depending on the structure of `rc` Attention this can of course be added again.

albertz commented on nn/transformer.py:
```diff
@@ -0,0 +1,220 @@
+  """
+  Defines one layer of a standard transformer encoder
+  """
+  def __init__(self, d_model, nhead, dim_feedforward=2048, drop=0.1, act=relu, layer_norm_eps=1e-5,
```
Types should be specified.

The param names are a bit inconsistent to each other. `d` vs `dim`. `nhead` vs `n_head` vs `num_head`.

Atticus1806 replied:
I updated a bunch of typing where I deemed it reasonable. 
One question regarding typing: For the activation functions, I put `activation` as a type since in the end `relu` etc. just "wrap" `activation`. Does that work for automatic typing checks or will that confuse the system?

albertz commented:
I see that the PyTorch argument names are somewhat inconsistent. The question is, should we adopt this (to make it very easy to maybe adopt other PyTorch code for RETURNN), or clean this up?

albertz commented on nn/transformer.py:
```diff
@@ -0,0 +1,220 @@
+from __future__ import annotations
+from .base import Module, LayerRef
+from .math_ import relu
+from ._generated_layers import Linear, layer_norm
```
`_generated_layers`, `math_`, `utils`, `container`, etc are kind of implementation details which are not really necessary here. Just import from `.`.

Atticus1806 commented:
>I see that the PyTorch argument names are somewhat inconsistent. The question is, should we adopt this (to make it very easy to maybe adopt other PyTorch code for RETURNN), or clean this up?

Maybe clean up and put the PyTorch into the docstring of the variable?

albertz commented:
> > I see that the PyTorch argument names are somewhat inconsistent. The question is, should we adopt this (to make it very easy to maybe adopt other PyTorch code for RETURNN), or clean this up?
> 
> Maybe clean up and put the PyTorch into the docstring of the variable?

Yes that's a good idea.

albertz commented on nn/transformer.py:
```diff
@@ -0,0 +1,219 @@
+  """
+  Defines one layer of a standard transformer encoder
+  """
+  def __init__(self, d_model: int, nhead: int, dim_feedforward: int = 2048, drop: float = 0.1, act: activation = relu,
```
The arguments are still not consistent.

The type of `act` is wrong.

Atticus1806 commented:
Okay so for now I decided on the following conventions: 

`d_model` becomes `dim_model` `dim_feedforward` becomes `dim_ff`. I first had `d_` prefix in mind, since they use it in Attention is all you need, but I am not a big fan of one letter prefixes, they are not always unambiguous. Shortening `feedforward` also seemed reasonable, here `ff` is what they also use in Attention is all you need.

For the `n` vs num` prefix I decided for `num`, since in `RETURNN` we also have `n_out` and I found it less confusing reserving `n` for that. Also again, one letter prefixes are somewhat weird in my opinion. Thus `n_head` becomes `num_heads`, `num_layers` stays the same. 

Next point `normalization` vs `norm` is now `norm`, shorter and more clear I think, this was initially not possible because for some reason I additionally imported `norm` from `.`, but this is gone now so this works fine.

`act` is now `activation` again, since its not imported aswell. This unfortunately does not work for `dropout`. So I think this has to stay `drop` or some variant of it.

I also added a bunch of documentation to the parameters. Let me know if I missed anything or you would name variables differently.

albertz commented:
> `d_model` becomes `dim_model`, `dim_feedforward` becomes `dim_ff`.
> `n_head` becomes `num_heads`, `num_layers` stays the same.
> Next point `normalization` vs `norm` is now `norm`
> `act` is now `activation` again

Yes, all good.

> this was initially not possible because for some reason I additionally imported `norm` from `.`
> This unfortunately does not work for `dropout`.

I don't understand this. Why is it not possible?

Atticus1806 commented:
>I don't understand this. Why is it not possible?

Because our function is also named `dropout`. I could import it with a different name but that would then just shift the issue in my opinion. If we use the same name this shadows right? Or I could name the var `dropout_val.

JackTemaki commented on nn/transformer.py:
```diff
@@ -0,0 +1,231 @@
+    :param norm_first: Whether to do layer norm before or afterwards
+    """
+    super().__init__()
+    self.self_attn = MultiheadAttention(dim_model, num_heads, dropout=dropout)  # will change with Attention Modules
```
Here you are passing the `dropout` function instead of a value. I think shadowing the global `dropout` with a local `dropout` in the constructor can be ok...

I am not sure if in general I would prefer to use also the `nn` module for internal code. Then errors like this do not happen.

albertz replied:
Yes, just using the `nn.` prefix is probably cleaner anyway.

Atticus1806 replied:
Yes good point, will change that

albertz commented:
> > I don't understand this. Why is it not possible?
> 
> Because our function is also named `dropout`. If we use the same name this shadows right?

I still don't understand. What's the problem with shadowing the global `dropout`?
And why do you anyway import `dropout` globally in this module?

JackTemaki commented on nn/transformer.py:
```diff
@@ -0,0 +1,231 @@
+  """
+  Defines the full Encoder of the standard transformer
+  """
+  def __init__(self, encoder_layer: Union[TransformerEncoderLayer, Any], num_layers: int, norm=None,
```
`norm`: type missing

Atticus1806 replied:
This is somewhat similar to `activation`, is there a type we can give `norm` that is specific enough to be clear?

albertz replied:
Just again `Callable[[LayerBase], LayerBase]`?

Atticus1806 replied:
Yes that would be correct, but I was thinking wether that is too general, since in the end we want activation functions (or normalization functions) in these cases and not just any function right? But maybe this also isnt a problem that its that general?

albertz replied:
But the type is anyway exactly that.

albertz commented on nn/transformer.py:
```diff
@@ -0,0 +1,231 @@
+    :param norm_first: Whether to do layer norm before or afterwards
+    """
+    super().__init__()
+    self.self_attn = MultiheadAttention(dim_model, num_heads, dropout=dropout)  # will change with Attention Modules
```
I don't understand the comment.

Atticus1806 replied:
Yeah the norm_first comment now reading it is somewhat weird, will update it

albertz replied:
I'm referring to the comment in this line here.

Atticus1806 replied:
these comments are internal, that this line will most likely change once AttentionModules are implemented, depending on what they require as input etc.

albertz replied:
By "internal" you mean still work-in-progress? Then write that in the comment. (TODO or WIP or so)
Although I would just remove the comment. The code could maybe just work like this.

Atticus1806 replied:
MultiheadAttention is just not implemented yet, so it is not clear what the input for that would be right now. I replaced the comment by: "to be implemented".

JackTemaki commented on nn/transformer.py:
```diff
@@ -0,0 +1,231 @@
+      """
+      :param decoder_layer: Decoder layer to be stacked num_layers times
+      :param num_layers: Number of layers
+      :param norm: normalization functional
```
Add to the docstring that this is only the final output norm.

albertz replied:
Why do you call it "functional"? Call it just "function".

albertz commented on nn/transformer.py:
```diff
@@ -0,0 +1,231 @@
+    :param num_heads: number heads, PyTorch name: nhead
+    :param dim_ff: dimension of feedforward layer, PyTorch name: dim_feedforward
+    :param drop: Dropout value, PyTorch name: dropout
+    :param activation: activation functional
```
```suggestion
    :param activation: activation function
```

albertz commented on nn/transformer.py:
```diff
@@ -0,0 +1,231 @@
+
+    return x
+
+  # self-attention block
+  def _sa_block(self, x: LayerRef) -> LayerRef:
```
Do not describe a function by a prefixed comment. Use docstrings for that.
But anyway, if you need to describe it because the name might be ambiguous, why don't you simply extend the name such that it is clear?
```suggestion
  def _self_attention_block(self, x: LayerRef) -> LayerRef:
```

Atticus1806 replied:
The name + comment directly come from PyTorch, but changing the name is fine too.

albertz replied:
We don't need to take it over exactly char by char.

albertz commented on nn/transformer.py:
```diff
@@ -0,0 +1,231 @@
+    return dropout(x, self.dropout)
+
+  # feed forward block
+  def _ff_block(self, x: LayerRef) -> LayerRef:
```
Same here.

albertz commented on nn/transformer.py:
```diff
@@ -0,0 +1,231 @@
+
+  # feed forward block
+  def _ff_block(self, x: LayerRef) -> LayerRef:
+    x = self.linear2(dropout(self.activationivation(self.linear1(x)), self.dropout))
```
Typo.

albertz commented on nn/transformer.py:
```diff
@@ -0,0 +1,231 @@
+    super().__init__()
+    self.self_attn = MultiheadAttention(dim_model, num_heads, dropout=dropout)  # will change with Attention Modules
+
+    self.linear1 = Linear(dim_ff)
+    self.linear2 = Linear(dim_model)
```
Use better names. Maybe like:
```suggestion
    self.linear_ff = Linear(dim_ff)
    self.linear_out = Linear(dim_model)
```

albertz commented:
Note that when we adopt #17, all the dimensions which are currently of type `int` would change the type to `DimensionTag`. Although the code otherwise would not change.

albertz commented on nn/transformer.py:
```diff
@@ -0,0 +1,231 @@
+
+from __future__ import annotations
+from typing import Optional, Any, Union
+from . import relu, layer_norm, dropout, LayerRef, Module, ModuleList, Linear
```
As discussed, do not do the global imports.
Instead just:
```suggestion
import . as nn
```

Atticus1806 replied:
My IDE does not recognize this.

Atticus1806 replied:
Error message is `identifier expected`

Atticus1806 replied:
importing `__init__` would work but this does not feel right either.

Atticus1806 replied:
`from .. import nn` would work

But regarding this: I would prefer import `LayerRef` at least and maybe also `Module` directly, since I feel like it makes the typing etc. harder to read with nn. prefix. Or would you not like that?

albertz replied:
But you get the idea. Then it's maybe `from .. import nn`.

albertz replied:
> it makes the typing etc. harder to read with nn. prefix

Why?

Atticus1806 replied:
Nvm  `from .. import nn would work` this also imports something else that is not the way we want it.

Atticus1806 replied:
Because with the `from ..` nn.relu is suddenly a tensorflow function. So what definitely works is `import returnn_common.nn as nn` but that would be a global import

>Why?

For me the `nn.` prefex is annoying to read, but its personal preference I guess, can also write it with the prefix.

Atticus1806 replied:
Hmm changed it for LayerRef to nn. locally and I guess its not as bad as I first thought. Nvm then.

So just need to agree on a proper way to import now. From my understanding you dont want global imports, I did not find a local way yet though, or at least didnt come up with the correct syntax.

albertz replied:
> Nvm `from .. import nn` would work this also imports something else that is not the way we want it.

I don't understand what you mean.

Atticus1806 replied:
>I don't understand what you mean.

The import does not import the `nn` submodule but something else (via `_lazy_loader`) which then does not find for example `nn.Module` and `nn.relu` references to a bunch of other functions which are not from `rc`.

albertz replied:
That sounds like a bug in `_lazy_loader`? Can't you just fix that?

Atticus1806 replied:
pushed the changes we accumulated this far now. I can have a try to look into it.

Atticus1806 replied:
So this seems to be a common problem in the interaction between IDE and `_lazy_loader`. 
I am now searching whether I can find a fix for this somewhere.

albertz replied:
I don't really understand.

Atticus1806 replied:
Okay I found the problem. The variable the lazy loader is saved in is called `nn` aswell, if we import from `..` this imports the variable instead of the module. Replacing the name by `_nn` (or some other name you prefer) will fix this.

albertz replied:
No, this is wrong. This is exactly the point. We want that. It should import the variable. But anyway, this does not belong here into this PR.

albertz replied:
I pushed some small change. Can you try whether this fixes the PyCharm issue?

albertz replied:
I pushed another change. Can you try again?

Atticus1806 replied:
yes it is working now, thank you!

albertz commented:
Btw, as usual, also see failing tests.

Atticus1806 commented:
>Btw, as usual, also see failing tests. 

But I think they are all to be expected, because the import does not properly work right now. (Besides the 3 for MultiheadAttention obviously).

albertz commented on nn/transformer.py:
```diff
@@ -0,0 +1,230 @@
+from __future__ import annotations
+from typing import Optional, Any, Union, Callable
+from .. import nn
+from . import LayerRef
```
Why do you have `LayerRef` as an explicit import? Why not just use `nn.LayerRef`? I think that would be more consistent.

Atticus1806 replied:
I was thinking, that in longer function heads this might become more messy and for typing I felt like this was fine to do, since like this we do have a clear separation between typing and actually using the class. e.g. When `LayerRef` would be used in code it would be via `nn.LayerRef` in the way I am thinking right now.

albertz replied:
But the difference between `LayerRef` and `nn.LayerRef` is really minor. Also, it should be consistent. Esp it should not be different for typing than for other code.

albertz commented on nn/transformer.py:
```diff
@@ -0,0 +1,230 @@
+    return nn.dropout(x, self.dropout)
+
+  def _feed_forward_block(self, x: LayerRef) -> LayerRef:
+    x = self.linear_out(nn.dropout(self.activation(self.linear_ff(x)), self.dropout))
```
My personal preference is (note the changed order):
```suggestion
    x = nn.Sequential(
      self.linear_ff,
      self.activation,
      functools.partial(nn.dropout, dropout=self.dropout),
      self.linear_out)(x)
```
Or:
```suggestion
    x = self.linear_ff(x)
    x = self.activation(x)
    x = nn.dropout(x, dropout=self.dropout)
    x = self.linear_out(x)
```

albertz commented on nn/transformer.py:
```diff
@@ -0,0 +1,230 @@
+
+  def _feed_forward_block(self, x: LayerRef) -> LayerRef:
+    x = self.linear_out(nn.dropout(self.activation(self.linear_ff(x)), self.dropout))
+    return nn.dropout(x, self.dropout)
```
Ah sorry, including that one, so only this here:
```suggestion
    return x
```

albertz commented on __init__.py:
```diff
@@ -22,4 +22,4 @@
   from . import nn  # noqa
 
 else:
-  nn = _LazyLoader("nn", globals())
```
First of all, whatever you try to fix/change here should be a separate PR.

Then further, this is obviously wrong. But let's discuss that in the separate PR.

Atticus1806 replied:
Hmm I think I see why this might be wrong, but then I am not really sure how else to fix it.

albertz commented on nn/transformer.py:
```diff
@@ -0,0 +1,237 @@
+
+    self.activation = activation
+    self.norm_first = norm_first
+    self.norm_eps = layer_norm_eps
```
Why is this not consistent?

albertz replied:
`layer_norm_eps` vs `norm_esp`

Atticus1806 replied:
yes you are right, changing to layer norm, since we always want to do (some) layer norm here.

albertz replied:
But see my other comments. Why do you have `norm` as an option in other places? This is also inconsistent. Either have it configurable everywhere in the same way, or nowhere.
Also, in any case, I would probably prefer `norm_eps`.

Atticus1806 replied:
Ahh you mean why it is an option in TransformerEncoder/Decoder but not in the layer itself?

Atticus1806 replied:
There is no reason, I can add that. Also If you prefer `norm` instead of `layer_norm` then we cut the `layer_prefix` from the variables.

albertz commented on nn/transformer.py:
```diff
@@ -0,0 +1,237 @@
+      """
+      super(TransformerDecoder, self).__init__()
+      import copy
+      self.layers = nn.ModuleList([copy.deepcopy(decoder_layer) for _ in range(num_layers)])
```
Better `Sequential`?

albertz replied:
Ah sorry, no this would not work for this case because `decoder_layer` requires two arguments.

Atticus1806 replied:
So this is resolved?

albertz commented on nn/transformer.py:
```diff
@@ -0,0 +1,237 @@
+    return x
+
+  def _self_attention_block(self, x: nn.LayerRef) -> nn.LayerRef:
+    x = self.self_attn(x, x, x)  # will change with Attention Modules
```
I don't understand the comment?

albertz commented on nn/transformer.py:
```diff
@@ -0,0 +1,237 @@
+    return nn.dropout(x, self.dropout)
+
+  def _multi_head_attention_block(self, x: nn.LayerRef, mem: nn.LayerRef) -> nn.LayerRef:
+    x = self.multihead_attn(x, mem, mem)  # will change with Attention Modules
```
Comment?

albertz commented on nn/transformer.py:
```diff
@@ -0,0 +1,237 @@
+    """
+    super().__init__()
+    self.self_attn = MultiheadAttention(dim_model, num_heads, dropout=dropout)  # to be implemented
+    self.multihead_attn = MultiheadAttention(dim_model, num_heads, dropout=dropout)  # to be implemented
```
You should not put such comment here. Either use "TODO" or don't put it at all here.

albertz commented on nn/transformer.py:
```diff
@@ -0,0 +1,237 @@
+    :param encoder_layer: Encoder layer to be stacked num_layers times
+    :param num_layers: Number of layers
+    :param norm: normalization function
+    :param layer_norm_eps: Epsilon value for layer normalization
```
Inconsistent to `norm`. Or is this specific to layer norm?

albertz replied:
It's still inconsistent.

Atticus1806 replied:
I changed it locally, marking it as done so I know what I have to still look at.

albertz replied:
No, push it first and then mark it as done.
But in any case, I guess you changed it to `norm_eps` now, right?

Atticus1806 replied:
Ah, ok. 
I changed the `layer_norm` prefix to `norm` everywhere it was used. So the normalization (and its parameters) now all have `norm` prefix.

albertz commented on nn/transformer.py:
```diff
@@ -0,0 +1,237 @@
+      :param decoder_layer: Decoder layer to be stacked num_layers times
+      :param num_layers: Number of layers
+      :param norm: normalization function for output layer normalization
+      :param layer_norm_eps: Epsilon value for output layer normalization
```
Inconsistent to `norm`. Or is this specific to layer norm?

albertz commented on nn/transformer.py:
```diff
@@ -0,0 +1,237 @@
+      """
+      :param decoder_layer: Decoder layer to be stacked num_layers times
+      :param num_layers: Number of layers
+      :param norm: normalization function for output layer normalization
```
Why do you have that option here for the decoder, but not for the encoder?

Atticus1806 replied:
Hmm what do you mean? norm is also a paramter for encoder.

albertz commented on nn/transformer.py:
```diff
@@ -0,0 +1,237 @@
+    :param dim_ff: dimension of feedforward layer, PyTorch name: dim_feedforward
+    :param dropout: Dropout value, PyTorch name: dropout
+    :param activation: activation function
+    :param layer_norm_eps: Epsilon value for layer normalization
```
Naming is inconsistent to `norm_first`.

albertz commented on nn/transformer.py:
```diff
@@ -0,0 +1,237 @@
+
+class TransformerDecoderLayer(nn.Module):
+  """
+  Defines one layer of a standard transformer decoder
```
Also describe the expected input shape and output shape.

albertz commented on nn/transformer.py:
```diff
@@ -0,0 +1,237 @@
+    """
+    super().__init__()
+    import copy
+    self.layers = nn.ModuleList([copy.deepcopy(encoder_layer) for _ in range(num_layers)])
```
I'm curious whether `deepcopy` works for our modules...
(Of course, it should, otherwise it would be a bug, and we should fix it.)

Atticus1806 replied:
So testing with a simple `Linear` this does not throw an error and yields another `Linear`.

Atticus1806 replied:
Not sure if this is what you wanted but I did a small test: 
```
layers = nn.Sequential([copy.deepcopy(nn.Linear(1)) for _ in range(3)])
nn.make_root_net_dict(layers,"data")
```
which then produces:
```
    {'0': {'class': 'linear', 'from': 'data:data', 'n_out': 1}, 
     '1': {'class': 'linear', 'from': '0', 'n_out': 1}, 
     '2': {'class': 'linear', 'from': '1', 'n_out': 1}, 
     'output': {'class': 'copy', 'from': '2'}}
 ```
So I am assuming `deepcopy` is working correctly from this.

albertz replied:
Looks good. You should make a small test case for this. In another separate PR. And remove the `[` after `nn.Sequential(`.

albertz commented on nn/transformer.py:
```diff
@@ -0,0 +1,237 @@
+    :param dropout: Dropout value, PyTorch name: dropout
+    :param activation: activation function
+    :param layer_norm_eps: Epsilon value for layer normalization
+    :param norm_first: if ``True`` will perform normalization before other att and ff operations, otherwise after
```
`norm_first` is inconsistent to `layer_norm_eps`

albertz replied:
`norm` is not documented.

albertz commented on nn/transformer.py:
```diff
@@ -0,0 +1,237 @@
+    :param norm_first: if ``True`` will perform normalization before other att and ff operations, otherwise after
+    """
+    super().__init__()
+    self.self_attn = MultiheadAttention(dim_model, num_heads, dropout=dropout)  # to be implemented
```
Remove this comment, or add "TODO".

albertz commented on nn/transformer.py:
```diff
@@ -0,0 +1,242 @@
+               num_decoder_layers: int = 6, dim_ff: int = 2048, drop: float = 0.1,
+               activation: Callable[[nn.LayerRef], nn.LayerRef] = nn.relu,
+               custom_encoder: Optional[Any] = None, custom_decoder: Optional[Any] = None,
+               norm_eps: float = 1e-5) -> None:
```
Why is there no `norm` argument here?

Atticus1806 replied:
This is took on from PyTorch, do you think it would make sense here? Usually the standard Transformer uses the usual layer normalization right? But I can add it if you want.

albertz replied:
But this is inconsistent or not? Why is it configurable for all the other modules but not here?

Atticus1806 replied:
yes makes sense.

albertz commented on nn/transformer.py:
```diff
@@ -0,0 +1,242 @@
+      self.encoder = custom_encoder
+    else:
+      encoder_layer = TransformerEncoderLayer(dim_model, num_heads, dim_ff, drop, activation, norm_eps, False,
+                                              nn.layer_norm)
```
Never use hanging indents.

albertz commented on nn/transformer.py:
```diff
@@ -0,0 +1,242 @@
+    if custom_encoder is not None:
+      self.encoder = custom_encoder
+    else:
+      encoder_layer = TransformerEncoderLayer(dim_model, num_heads, dim_ff, drop, activation, norm_eps, False,
```
For such long argument list, always use named arguments.

albertz commented on nn/transformer.py:
```diff
@@ -0,0 +1,242 @@
+      self.decoder = custom_decoder
+    else:
+      decoder_layer = TransformerDecoderLayer(dim_model, num_heads, dim_ff, drop, activation, norm_eps, False,
+                                              nn.layer_norm)
```
Never use hanging indents.

Atticus1806 replied:
is it better like this?

albertz commented on nn/transformer.py:
```diff
@@ -0,0 +1,242 @@
+    :param dropout: Dropout value, PyTorch name: dropout
+    :param activation: activation function
+    :param norm_eps: Epsilon value for layer normalization
+    :param norm_first: if ``True`` will perform normalization before other att and ff operations, otherwise after
```
`norm` is not documented.

albertz commented on nn/transformer.py:
```diff
@@ -0,0 +1,242 @@
+  """
+  def __init__(self, dim_model: int, num_heads: int, dim_ff: int = 2048, dropout: float = 0.1,
+               activation: Callable[[nn.LayerRef], nn.LayerRef] = nn.relu, norm_eps: float = 1e-5,
+               norm_first: bool = False, norm: Callable[[nn.LayerRef], nn.LayerRef] = None):
```
If `norm` is non-optional here, it should not use the default `None` but `nn.layer_norm`.

Atticus1806 commented:
Also it seems like it does not like the typing for `norm` as it is. Do you have an idea what we could put there instead? Something that hints the typing that there is a second argument I guess.

albertz commented:
> Also it seems like it does not like the typing for `norm` as it is. Do you have an idea what we could put there instead? Something that hints the typing that there is a second argument I guess.

Yes, that is via [`typing.Protocol`](https://docs.python.org/3/library/typing.html#typing.Protocol), see [here](https://stackoverflow.com/questions/57837609/python-typing-signature-typing-callable-for-function-with-kwargs) or [here](https://mypy.readthedocs.io/en/stable/protocols.html#callback-protocols). Although `typing.Protocol` requires Python 3.8. Or there is the external package `typing_extensions` for earlier Python versions. Not sure if we want to add that.

albertz commented on nn/transformer.py:
```diff
@@ -0,0 +1,253 @@
+      decoder_layer = TransformerDecoderLayer(
+        dim_model=output_dim, num_heads=num_heads, dim_ff=dim_ff, dropout=dropout, activation=activation,
+        norm_eps=norm_eps, norm=norm
+      )
```
Do the closing bracket on the line before.

albertz commented on nn/transformer.py:
```diff
@@ -0,0 +1,253 @@
+      encoder_layer = TransformerEncoderLayer(
+        dim_model=output_dim, num_heads=num_heads, dim_ff=dim_ff, dropout=dropout, activation=activation,
+        norm_eps=norm_eps, norm=norm
+      )
```
Closing bracket on the line before.

albertz commented on nn/transformer.py:
```diff
@@ -0,0 +1,237 @@
+               activation: Callable[[nn.LayerRef], nn.LayerRef] = nn.relu, layer_norm_eps: float = 1e-5,
+               norm_first: bool = False):
+    """
+    :param dim_model: hidden dim, PyTorch name: d_model
```
This is still wrong.

Atticus1806 replied:
would you change it for all Layers? Changed it for the Transformer Module, but thought hidden would be fine for Layers inside the Transformer. But ofc. can change every of those, if you prefer it that way.

albertz replied:
No, the documentation on a module should just be about this module isolated, no matter where or how it is used (i.e. it is not relevant whether it can be used such that it is hidden).

albertz commented on nn/transformer.py:
```diff
@@ -0,0 +1,253 @@
+
+  def forward(self, tgt: nn.LayerRef, memory: nn.LayerRef) -> nn.LayerRef:
+    """
+    Executes the decoder layer as often as in num_layers defined
```
This comment is very misleading, or actually wrong.
- "Execute" is the wrong wrong term. It applies it.
- It does not apply the decoder layer multiple times, no. It applies all the decoder layer**s** once. Every decoder layer is a copy of the original decoder layer.

Atticus1806 commented:
>Not sure if we want to add that.

So maybe just change it to `Callable` without any parameters?

albertz commented:
> So maybe just change it to `Callable` without any parameters?

That's invalid. Just leave it away then.

albertz commented:
Note, I implemented already some initial variant of self-attention (#52). So we can already partly test this.

Still also code style test is failing.

albertz commented:
I don't see any positional encoding (relative, absolute, or any) in the code.

(Btw, also the original PyTorch Transformer lacks this. There is some interesting relevant on-going discussion here: https://github.com/pytorch/pytorch/issues/24826.)

(For reference, [Fairseq implementation](https://github.com/pytorch/fairseq/blob/0b21875e45f332bedbcc0617dcf9379d3c03855f/fairseq/modules/positional_embedding.py#L12).)

When we add it, I think we definitely need relative pos encoding. Not sure if abs pos enc is even needed, as rel pos is always better, right?

albertz commented:
Also interesting for reference is the original discussion on adding the Transformer to PyTorch: https://github.com/pytorch/pytorch/issues/10459

Atticus1806 commented:
So I did a rebase and used `nn.SelfAttention`. From what I can see the Attention for Decoder/Encoder Attention is not there yet right? Note that I am only now looking into `DimensionTags` in Detail and tried to fill the other parameters to the best of my knowledge but this might be wrong as it right now.

Atticus1806 commented:
>When we add it, I think we definitely need relative pos encoding. Not sure if abs pos enc is even needed, as rel pos is always better, right?

But doesn't it make sense to implement it separate from the `Transformer` to keep our modularity consistent? While its needed for the Transformer embedding and pos.enc. are something used also without the model, so I would vote for splitting this.

albertz commented:
> > When we add it, I think we definitely need relative pos encoding. Not sure if abs pos enc is even needed, as rel pos is always better, right?
> 
> But doesn't it make sense to implement it separate from the `Transformer` to keep our modularity consistent? While its needed for the Transformer embedding and pos.enc. are something used also without the model, so I would vote for splitting this.

Yes sure definitely. I was just saying, our Transformer module should be flexible enough such that this is very easy to use.

I didn't really thought much about how it would actually look like. Such an example would be good in any case.

Probably it can be made as a generic extension to the `SelfAttention` module, similar as the `SelfAttentionLayer` in RETURNN. Or maybe it could be a separate module. I don't know.

So the `Transformer` module should not hardcode the usage of `SelfAttention` but somehow make that configurable. Or we should just think about what would be the easiest and most straightfoward way for the user to use this.

Atticus1806 commented:
>So the Transformer module should not hardcode the usage of SelfAttention but somehow make that configurable. 

It is always possible to give a custom Encoder oder Decoder, so on a higher level this is possible already. Maybe we could extend the logic to check whether a custom Encoder/Decoder layer is given and use that as the Encoder/Decoder stack instead. So similar to giving a full custom Encoder/Decoder. That should be specific enough to also allow custom Attention mechanisms. I don't think I would actually make the Attention a parameter in the layer itself.

albertz commented:
> > So the Transformer module should not hardcode the usage of SelfAttention but somehow make that configurable.
> 
> ... I don't think I would actually make the Attention a parameter in the layer itself.

I'm not sure. Have you seen other relative positional encoding usages and code examples?

As usual, we probably should take a look at other frameworks, like Fairseq, ESPNet, Lingvo, etc.

albertz commented on nn/transformer.py:
```diff
@@ -0,0 +1,253 @@
+    super().__init__()
+    time_dim = nn.DimensionTag(nn.DimensionTag.Types.Spatial, description="time")
+    self.self_attn = nn.SelfAttention(
+      axis=time_dim, key_dim_total=dim_model, value_dim_total=dim_model, num_heads=num_heads, att_dropout=dropout)
```
`att_dropout=dropout` is wrong.

albertz replied:
Oh wait, actually not? It uses the same dropout also for the attention dropout? At least in the PyTorch implementation. But is this good? Does it make sense to use the same dropout value?

albertz replied:
Yes I would expect it to be separate. I would not expect that the same dropout value makes sense for both attention weights and for other activations.

What do you usually use? Or what is usually used in the literature? Is it common that people use the same dropout for both?

If this is very common and a good choice, then we can leave it. Otherwise I think we should separate it.

Atticus1806 replied:
At least in PyTorch its the same value I think, looking at Attention is all you need I am not sure if they even used dropout there. At least the paper does not explicitly state it from what I can see. But if you want we can separate and use the same defaults.

albertz replied:
Look in actual configs from publications (ours, others) or production systems or our baselines.

It also doesn't make sense to use the same defaults. The defaults should be good. That was what I asked before. What are common good defaults here?

Atticus1806 replied:
From what I could find 0.1 is the default for both attention dropout and (general) dropout, correlating to the 6 Dec/Enc Layer Transformer. For other sizes there are other defaults, but if we use 6 as default this should be standard in literature. I separated it from standard dropout now.

albertz replied:
Where? Esp what recent papers use that? Please add such references to the docstring as well.

albertz replied:
But also, what do we usually use? What do you have in your baselines?

Atticus1806 replied:
espnet uses 0.1 for both for TTS, for ASR Encoder/Decoder they use 0.1 dropout, 0.0 self attention dropout
fairseq uses 0.1 for dropout and 0.0 for self attention dropout
Lingvo uses 0.0 as a default
[ASR Transformer](https://ieeexplore.ieee.org/stamp/stamp.jsp?tp=&arnumber=8462506) uses 0.1 for both
for some other papers I could not find information on what they use or they referenced to the above paper, like the espnet [study](https://merl.com/publications/docs/TR2019-158.pdf)

Atticus1806 replied:
So my suggestion (if we want to fill it and not leave it empty) would be to use 0.1 for residual dropout and either 0.1 or 0.0 for attention dropout. 0.0 would match with your default from the `SelfAttention` layer.

albertz replied:
Maybe then 0.1 for both. I think it is a good choice to use it always also for the att weights.

Also, `SelfAttention` maybe should change the default.

Atticus1806 commented:
>I'm not sure. Have you seen other relative positional encoding usages and code examples?
As usual, we probably should take a look at other frameworks, like Fairseq, ESPNet, Lingvo, etc.

So I had a look and the they all hade some kind of encoding included. Fairseq fixed it while ESPNet and Lingvo gave a chance to give it as a Parameter (with decent default). So I decided on doing the same here, using `nn.positional_encoding` as standard but giving the chance to pass a different variation (or even explicitly setting it to None if no pos_enc is aimed for).

albertz commented on nn/transformer.py:
```diff
@@ -69,15 +69,17 @@ class TransformerEncoder(nn.Module):
   """
   def __init__(self, encoder_layer: Union[TransformerEncoderLayer, Any], num_layers: int,
-               norm=nn.layer_norm, norm_eps: float = 1e-5):
+               norm=nn.layer_norm, norm_eps: float = 1e-5, pos_enc: Callable = nn.positional_encoding):
```
`Callable` is the wrong type.

Atticus1806 replied:
What instead would you suggest? I have not really dealt with positional encodings in my work yet, so I dont really have good knowledge on what would be better. I tried to have a look at other Frameworks and tried to replicate it here.

albertz commented on nn/transformer.py:
```diff
@@ -0,0 +1,281 @@
+    :param custom_encoder: Custom Encoder to replace the standard encoder
+    :param custom_decoder: Custom Decoder to replace the standard decoder
+    :param custom_encoder_layer: Custom Encoder layer to replace the standard layer if custom_encoder and
+                                 custom_encoder_layer are given custom_encoder will be preferred
```
Do not indent like this.

albertz replied:
Always indent by 2 spaces.

albertz commented on nn/transformer.py:
```diff
@@ -0,0 +1,281 @@
+    :param custom_encoder_layer: Custom Encoder layer to replace the standard layer if custom_encoder and
+                                 custom_encoder_layer are given custom_encoder will be preferred
+    :param custom_decoder_layer: Custom Decoder layer to replace the standard layer if custom_decoder and
+                                 custom_decoder_layer are given custom_decoder will be preferred
```
Indent...

albertz commented:
> > I'm not sure. Have you seen other relative positional encoding usages and code examples?
> > As usual, we probably should take a look at other frameworks, like Fairseq, ESPNet, Lingvo, etc.
> 
> So I had a look and the they all hade some kind of encoding included. Fairseq fixed it while ESPNet and Lingvo gave a chance to give it as a Parameter (with decent default). So I decided on doing the same here, using `nn.positional_encoding` as standard but giving the chance to pass a different variation (or even explicitly setting it to None if no pos_enc is aimed for).

`nn.positional_encoding` is the absolute pos encoding (maybe we should rename it to make that more clear).

While that is what they used in the original paper, it is well known that this is not a good default choice. Relative pos encoding is always better. So we should not use this as the default. We should use rel pos encoding as the default.

albertz commented on nn/transformer.py:
```diff
@@ -11,7 +11,7 @@ class TransformerEncoderLayer(nn.Module):
   Defines one layer of a standard transformer encoder
   """
-  def __init__(self, dim_model: int, num_heads: int, dim_ff: int = 2048, dropout: float = 0.1,
+  def __init__(self, dim_model: int, num_heads: int, dim_ff: int = 2048, dropout: float = 0.1, att_dropout: float = 0.1,
```
Why do you use these defaults for `dropout` and `att_dropout`? Can you point to some recent published papers where they use the same, or some other references?

JackTemaki replied:
I would recommend to not put any defaults at all. ASR/MT and TTS can all have very different reasonable values (usually between 0.1 and 0.5).

albertz commented on nn/transformer.py:
```diff
@@ -0,0 +1,286 @@
+    :param norm_eps: Epsilon value for layer normalization
+    :param norm_first: if ``True`` will perform normalization before other att and ff operations, otherwise after
+    :param norm: normalization function
+    :param att_dropout: dropout value for attention
```
The order of the params in the docstring should be the same as in the function.

albertz commented on nn/transformer.py:
```diff
@@ -0,0 +1,286 @@
+        decoder_layer = TransformerDecoderLayer(
+          dim_model=output_dim, num_heads=num_heads, dim_ff=dim_ff, dropout=dropout, activation=activation,
+          norm_eps=norm_eps, norm=norm, att_dropout=att_dropout)
+      self.decoder = TransformerDecoder(
```
I wonder now what we should use here.

We can either define the whole `Transformer` step-wise (as `TransformerStep` or so) and then use `TransformerDecoderStep` here. But I wonder if this is useful.

Or we also define a `TransformerDecoder`. But what should it do? Use a ground truth input? But this only makes sense in training and not in recognition.

I think the `Transformer` should be useful for both cases. So there might be some `search` option to the module which defines whether it uses a ground truth target or uses search. Or some other way to determine this (see #18).

We should maybe abstract that away, i.e. a generic interface from the decoder output to some `choice` module in search, and using the prev output of that as input to an embedding, and in training using the ground truth as input. Although that should be a bit more generic, also covering the regression case (TTS).

Atticus1806 replied:
How would you initialized the state for the TransformerDecoderStep? 
So was thinking in Transformer something like: 
```
  def forward(self, source: nn.LayerRef, target: nn.LayerRef) -> nn.LayerRef:
    """
    Forward step of Transformer
    """
    memory = self.encoder(source)
    with nn.Loop() as loop:
      target_ = loop.unstack(target, axis="T")
      memory_ = loop.unstack(memory, axis="T")
      loop.state.decoder = nn.State(initial=0)
      output_, state = self.decoder(target_, memory=memory_, state=loop.state.decoder)
      output = loop.stack(output_)

    return output
```
But I am not sure about the `loop.state.decoder` line, do we have to initalize each of the recurrent parameters by hand? so in this case `k_accum` and `v_accum` or also others I am missing right now? Is the Stacking/Unstacking even correct?

albertz replied:
`memory` has a different axis, and you would not unstack that put pass it as a whole.

`target` is only available during training. Or more specific, it must be provided when you don't use search, but with search it would use the previous predicted. How we define `search` is also still open (#18). Maybe just a `search: bool` option to the module.

For recurrent modules, we have the convention that they should provide an `initial_state` function. So you would do:
```
loop.state.decoder = nn.State(initial=self.decoder_initial_state())
```

The return value of `self.decoder(...)` would also be assigned again to `loop.state.decoder`.

The target labels also need some state because you must access the previous value.
Also you would return the last state. And it should accept the initial state.
So:
```
def forward(..., initial_state: nn.LayerState) -> Tuple[nn.LayerRef, nn.LayerState]:
  ...
    loop.state = initial_state
    logits, loop.state.decoder = self.decoder(loop.state.target, memory=memory, state=loop.state.decoder)
    if self.search:
      loop.state.target = nn.choice(logits, input_type="logits")
    else:
      loop.state.target = loop.unstack(target)
    outputs = loop.stack(loop.state.target)
  return outputs, loop.state
```
(Just a draft. I think there are multiple things here which still would not work correctly like that.)

albertz replied:
And the `initial_state` of `Transformer` would look sth like this:
```
def initial_state(initial_target: nn.LayerRef = 0) -> nn.LayerState:
  return nn.LayerState(
    target=initial_target,
    decoder=self.decoder.initial_state())
```

And `initial_state` of the `TransformerDecoderStep` would collect the initial state of all the layers.

And `initial_state` of each `TransformerDecoderLayerStep` would use the `initial_state` of the `SelfAttentionStep`.
(Which is also missing, but this will be implemented.)

Atticus1806 replied:
I think I get what you mean. I will code a draft and push it, so we can then from there further refine it.

albertz commented:
Here something about the relative pos encoding (specifically in Transformer XL, which is a bit extended from the original rel pos enc, but maybe better, also the default in Conformer): https://lilianweng.github.io/lil-log/2020/04/07/the-transformer-family.html#longer-attention-span-transformer-xl

Atticus1806 commented:
Okay I pushed an updated version including: 
States for Decoder Steps
initial states for the layers (even though I am not sure about the `DecoderStep` init, this might be wrong.
First implementation of the Transformer Decoder Loop with search flag and beamsize

albertz commented on nn/transformer.py:
```diff
@@ -0,0 +1,320 @@
+    """
+    initial state declaration
+    """
+    return nn.LayerState({mod: mod.initial_state() for mod in self.layers})
```
```suggestion
    return nn.LayerState({key: mod.initial_state() for mod in self.layers.named_children()})
```

albertz commented on nn/transformer.py:
```diff
@@ -0,0 +1,320 @@
+    :param encoder_pos_enc: Positional Encoding for the Encoder
+    :param decoder_pos_enc: Positional Encoding for the Decoder
+    :param search: Whether search is enabled or not
+    :param beam_size: Beam Size for search
```
beam_size and other search specific things should not be added here. This should somehow be more generic. Either via #18 or #49.

albertz commented on nn/transformer.py:
```diff
@@ -0,0 +1,320 @@
+    self.search = search
+    if search:
+      assert beam_size is not None
+      self.beam_size = beam_size
```
You never should only conditionally assign an attrib. An attrib should always be there. Set `self.beam_size = None` otherwise or so.

But anyway, as explained before, beam size should not be here.

albertz commented on nn/transformer.py:
```diff
@@ -0,0 +1,320 @@
+      assert beam_size is not None
+      self.beam_size = beam_size
+
+  def forward(self, source: nn.LayerRef, target: nn.LayerRef,
+              initial_state: nn.LayerState) -> Tuple[nn.LayerRef, nn.LayerState]:
```
The other args despite `source` should be keyword-only. So you miss a `, *,` after `source`.

`target` is optional, as explained before.

`initial_state` is also optional. It would use `self.initial_state()` if not given.
```suggestion
  def forward(self, source: nn.LayerRef, *, target: Optional[nn.LayerRef] = None,
              initial_state: Optional[nn.LayerState] = None) -> Tuple[nn.LayerRef, nn.LayerState]:
```

Atticus1806 commented on nn/transformer.py:
```diff
@@ -0,0 +1,320 @@
+    :param norm_eps: Epsilon value for layer normalization
+    :param norm: function for layer normalization
+    :param encoder_pos_enc: Positional Encoding for the Encoder
+    :param decoder_pos_enc: Positional Encoding for the Decoder
```
Yes that is what I am looking into next, but I guess its better to remove the "old" encoding for now.

albertz commented on nn/transformer.py:
```diff
@@ -0,0 +1,301 @@
+    Two possible forward variants of decoder, defined by self.norm_first, x and memory have shape {B, T, F}
+    """
+    if self.norm_first:
+      x, new_state = x + self._self_attention_block(self.norm(x, epsilon=self.norm_eps), state=state)
```
This is wrong. I guess `new_state` comes from `_self_attention_block`? Then you cannot add `x` to the tuple.

albertz commented on nn/transformer.py:
```diff
@@ -0,0 +1,301 @@
+      x = x + self._multi_head_attention_block(self.norm(x, epsilon=self.norm_eps), memory)
+      x = x + self._feed_forward_block(self.norm(x, epsilon=self.norm_eps))
+    else:
+      x, new_state = self.norm(x + self._self_attention_block(x, state=state), epsilon=self.norm_eps)
```
Same problem here.

albertz commented on nn/transformer.py:
```diff
@@ -0,0 +1,301 @@
+    initial state declaration
+    """
+    return nn.LayerState(
+      self_attention=self.self_attn.initial_state(),
```
Always check consistency. If the attrib is called `self_attn`, you also should call the state key the same, i.e. `self_attn`.

albertz commented on nn/transformer.py:
```diff
@@ -0,0 +1,301 @@
+    """
+    return nn.LayerState(
+      self_attention=self.self_attn.initial_state(),
+      attention=self.multihead_attn.initial_state())
```
Same here. Although I would anyway also rename `multihead_attn`.

albertz commented on nn/transformer.py:
```diff
@@ -0,0 +1,301 @@
+    super().__init__()
+    self.self_attn = nn.SelfAttentionStep(
+      key_dim_total=dim_model, value_dim_total=dim_model, num_heads=num_heads, att_dropout=att_dropout)
+    self.multihead_attn = nn.Attention()
```
Keep it consistent.
```suggestion
    self.attn = nn.Attention()
```

Despite, maybe instead of using `attn` (here and before), maybe `attention` instead?

albertz commented on nn/transformer.py:
```diff
@@ -0,0 +1,301 @@
+  """
+  def __init__(self, decoder_layer: Union[TransformerDecoderLayerStep, Any], num_layers: int,
+               norm=nn.layer_norm, norm_eps: float = 1e-5):
+      """
```
Indentation is wrong (4 spaces, should be 2).

albertz commented on nn/transformer.py:
```diff
@@ -0,0 +1,301 @@
+    """
+    output = x
+
+    for mod in self.layers:
+      output, state = mod(output, memory=memory, state=state)
```
This is wrong. You don't pass the same state through all the modules.
Every module has an own state.
```suggestion
    for key, mod in self.layers.named_children():
      output, state[key] = mod(output, memory=memory, state=state[key])
```

albertz commented on nn/transformer.py:
```diff
@@ -0,0 +1,301 @@
+    """
+    Forward step of Transformer
+    """
+    memory = self.encoder(source)
```
You also need sth like:
```
if not initial_state:
  initial_state = self.initial_state()
```

patrick-wilken commented:
Looks good on a high level.

Parameter initialization could be different from what we use. It is not set here, so the default `glorot_uniform` of `LinearLayer` is used, right? We used to have `variance_scaling_initializer(mode='fan_in', distribution='uniform', scale=0.78)` in our config, but switched some time ago to `variance_scaling_initializer(mode='fan_avg', distribution='uniform', scale=scale)`, where `scale` is 0.5 for some layers, 1 for others, copying what is done in fairseq. I would have to dig more into this to understand the differences, but Tamer reported that this improved training quite a bit.

Also, the `norm_eps=1e-5` is different from the `LayerNormLayer` default of `1e-6` which we use, not sure what's the effect of this.

albertz commented:
Note on param init, it's still kind of open how we handle that here (#59).

albertz commented:
@patrick-wilken Can you maybe comment on other aspects? What positional encoding is currently the standard or best? Do you still use vanilla Transformer (just as defined here) or are there other things?

patrick-wilken commented:
Yes, vanilla transformer with what would be `norm_first=True` here and absolute positional embeddings. Relative positional embeddings seem to be better for long, unseen sequence lengths ([see here](https://www-i6.informatik.rwth-aachen.de/publications/download/1132/RosendahlJanTranVietAnhKhoaWangWeiyueNeyHermann--AnalysisofPositionalEncodingsforNeuralMachineTranslation--2019.pdf)), but they are not widely used for standard MT. I couldn't find an implementation in fairseq and OpenNMT, for example.

albertz commented:
Isn't sth like transformer XL becoming the standard? I thought that
document level translation is also more standard now.


Patrick Wilken ***@***.***> schrieb am Fr., 12. Nov. 2021,
17:18:

> Yes, vanilla transformer with what would be norm_first=True here and
> absolute positional embeddings. Relative positional embeddings seem to be
> better for long, unseen sequence lengths (see here
> <https://www-i6.informatik.rwth-aachen.de/publications/download/1132/RosendahlJanTranVietAnhKhoaWangWeiyueNeyHermann--AnalysisofPositionalEncodingsforNeuralMachineTranslation--2019.pdf>),
> but they are not widely used for standard MT. I couldn't find an
> implementation in fairseq and OpenNMT, for example.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/rwth-i6/returnn_common/pull/55#issuecomment-967239427>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAAON7FIDPVDALCXMEEN4JTULU455ANCNFSM5HIWQFMA>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
>

Atticus1806 commented:
> Yes, vanilla transformer with what would be norm_first=True here and absolute positional embeddings.

So would you rather have `True` as default for `norm_first` ? 

> Also, the norm_eps=1e-5 is different from the LayerNormLayer default of 1e-6 which we use, not sure what's the effect of this.

I just double checked with the implementation of Attention Is All You need and it seems they use 1e-6 aswell. I am not sure why PyTorch hast 1e-5 there, maybe its one of the inconsistencies.

albertz commented:
> > Yes, vanilla transformer with what would be norm_first=True here and absolute positional embeddings.
> 
> So would you rather have `True` as default for `norm_first` ?

The original Transformer is actually with `norm_first=False` ([paper](https://arxiv.org/pdf/1706.03762v1.pdf): `LayerNorm(x + Sublayer(x))`).

We should definitely not just follow the original paper. It is very well known that this is one aspect where it was clearly suboptimal.

In Tensor2Tensor (which is kind of the official implementation), you find this ([here](https://github.com/tensorflow/tensor2tensor/blob/86caf01fd174c2ea2d3c789abb27a43f8948df1f/tensor2tensor/layers/common_hparams.py#L127)):
```
# Sequences of operations to perform on layer input and layer output.
# Used by common_layers.layer_preprocess, common_layers.layer_postprocess
# Each character represents an operation:
# none: no preprocessing
#    d: apply dropout
#    n: apply normalization (see norm_type and norm_epsilon)
#    a: add layer input (residual connection - only during postprocess)
# The special string "none" is used instead of the empty string
# to indicate no pre/postprocessing, since the empty string causes
# trouble for hyperparameter tuning.
# TODO(noam): The current settings ("", "dan") are the published version
# of the transformer.  ("n", "da") seems better for harder-to-learn
# models, so it should probably be the default.
layer_preprocess_sequence="none",
layer_postprocess_sequence="dan",
```

So, I tend to think as well that `norm_first=True` is maybe a better default.

> > Also, the norm_eps=1e-5 is different from the LayerNormLayer default of 1e-6 which we use, not sure what's the effect of this.
> 
> I just double checked with the implementation of Attention Is All You need and it seems they use 1e-6 aswell. I am not sure why PyTorch hast 1e-5 there, maybe its one of the inconsistencies.

Again, I don't understand the reasoning. The original Transformer settings are not optimal and better ones are known. PyTorch defaults are maybe better than what is in the original paper. But as I said, you should really check recent relevant literature.

Atticus1806 commented:
> But as I said, you should really check recent relevant literature.

Okay so I checked: fairseq has `1e-5` as default for the `LayerNorm` layer, from what I can see they dont change it in their default transformer. Lingvo uses `1e-6`. espnet uses as a default in the `LayerNorm` layer 1e-12 which seems weirdly off from the e-5 or e-6 but i could not find where they might change that, it seems that for the standard transformer they use that as a value. Apart form these values I found it hard to find a paper which talks about the actual impact of the value. All I could find were talking about the position or the type of normalization, but I could not find an analysis about possible values. 

Regarding relative positional encodings: I did quite some reading into (relative) positional encodings trying to understand how they work and how to implement them. If I understood correctly, this is part of the `SelfAttention` calculation. What do you think about leaving positional encodings out of this model for now? When doing absolute encodings this can be done before the Transformer and the PR implementing relative positional encodings for SelfAttention could add a flag to the transformer whether to enable it or not. Ofcourse we can also agree on an interface beforehand and already include that flag in this PR.

albertz commented:
> What do you think about leaving positional encodings out of this model for now?

As relative positional encoding is some property deep inside the Transformer (inside the self attention), and also an important mechanism at least for certain tasks, I think we should not ignore this now, because it might have a big influence on our design. Our design should allow to make use of relative pos encoding in an easy way.

Or more generally: The current Transformer design should allow for changes insides the core self attention process in an easy way. E.g. also to replace the self attention by other attention types, like sparse attention, using LHS and all that, Linformer, etc. (@Zettelkasten)

I think the current design does not really allow this. The way you would currently do that is probably to copy & paste the current Transformer code and replace the self attention by some own custom thing. Which basically shows that it does not allow for that. (Note that this is maybe not always bad. In some situations, this approach could just be fine.)

However, whatever potential generic API we propose here, we should also be careful that it is not too difficult to follow, because it is too abstract, has too much indirections, etc. Or even worse, having some quite abstract API, which allows for rel pos encoding, but it would turn out later that it is not really generic enough to allow other things, so it is really only useful for this one specific thing, and thus pointless (if we want it to be specific, it could just be a flag `rel_pos_encoding: bool` anyway, but we want to avoid that).

Maybe this is tricky to get right. Or maybe not really possible. But we should at least think about it.

Atticus1806 commented:
> The current Transformer design should allow for changes insides the core self attention process in an easy way. 

Trying to phrase it more general: The transformer is a stack of multiple components. Depending on view and features this amount can be a bit different but in general we have smth. like a potential `PreEncoderNet` where e.g. (if used) absolute encodings are applied, an `EncoderNet`, a `PreDecoderNet`, a `DecoderNet` and a `PostNet`. The Encoder and Decoder Nets just stack Encoder/Decoder layer ontop of each other in most of the cases.
The point I am getting to is: I am wondering whether we should maybe provide 2 variations: One "Plug and Play" Transformer with standard architecture (possibly including improvements like relative encodings etc.), solid defaults and features that are required by the users for the vanilla transformer, but at the price of flexibility and then a second, lets call it for now, `TransformerConstructor` which gets as input possibly there building blocks I talked about before, potentially then stacking Encoder and Decoder layers, but other than that mainly handles the merging together of these components. What do you think?

albertz commented:
When I hear "Plug and Play", my association is a flexible set of building blocks which can be plugged together in many ways, so basically what you describe in your second variant. But anyway, this is just terminology.

I'm not sure if we really need a `TransformerConstructor` class or so. Maybe the building blocks are already enough.

We should make this a bit more concrete. How do the building blocks look like? What is different from what we already have right now? Because we already have building blocks right now. They are just designed in a kind of hierarchical way. `Transformer` uses `TransformerEncoder`, `TransformerEncoder` uses `TransformerEncoderLayer`, and `TransformerEncoderLayer` uses `SelfAttention`. So when you want to replace or modify `SelfAttention` in this design, you would need to basically rebuild everything.

I don't really have a good solution so far. I have some ideas but I'm not exactly happy with them. I think we should think a bit more on this.

In any case, yes, there should be a ready-to-use Transformer as well. But this should then just be based on the building blocks, and should not be another separate implementation.

Also, this ready-to-use Transformer should in any case have an easy way to use relative positional encoding. Maybe that's not always needed, but this will be often needed. Maybe that should even be the default. Maybe specifically the Transformer XL variant.

albertz commented:
We maybe should also not overthink this. It doesn't need to be too generic. What we want is an easy way to replace the default self attention inside the Transformer, as this is probably some frequent thing to change. So maybe `self_att` can just be an argument of the `Transformer` module, or maybe separately `enc_self_att` and `dec_self_att`.

Atticus1806 commented:
>So maybe self_att can just be an argument of the Transformer module, or maybe separately enc_self_att and dec_self_att.

Would you hand over a initialized layer which just is called in forwad or would you init it within the modules? Usually I would prefer within the modules, but then the arguments are fixed.

albertz commented:
> > So maybe self_att can just be an argument of the Transformer module, or maybe separately enc_self_att and dec_self_att.
> 
> Would you hand over a initialized layer which just is called in forwad or would you init it within the modules? Usually I would prefer within the modules, but then the arguments are fixed.

Why would you init within the modules?

Of course, we need the ability here to pass in any module with any options. There are many options, like:

- Passing `module` (instance of `Module`). It would be `deepcopy`d just like we do e.g. for `TransformerEncoderLayer`. So this is kind of consistent to other existing code.
- Passing `module_cls`, `module_args`, `module_kwargs`. Ugly in my opinion.
- Passing `module_constructor` (or so), which is supposed to be a function without arguments. So the user could pass sth like `lambda: SelfAttention(...)`. This is also very flexible but also maybe a bit ugly and maybe unintuitive.

Atticus1806 commented:
Decided for your first option for now.  Also changed `norm_first` default to `True` since I understood this should be better. 
Would you type them with `nn.Module` ? This causes Pycharm to say Module does not have an inital state function, but not sure if this is relevant. 

So in general what is "missing" for this to be finished is: 

- `nn.SelfAttention` initial state
- `nn.Attention` implementation, I think this should be an own module.
- `self.search`
- positional encodings, still not sure if I understand correctly, but should this be part of `nn.SelfAttention`? From my understanding its a trainable weight added during self attention right?

Atticus1806 commented:
> Edit: Or is this working? If so, how?

Removed all init parts of search for now only kept it in forward since we agreed on that api beforehand.

albertz commented:
See also this: https://github.com/pytorch/pytorch/issues/67999

albertz commented:
I think we should not try to get this here into a perfect and ready state but just merge it soon and then iterate on it.

@Atticus1806 When do you think would be a good time to merge? Do you want to improve anything further before the merge, or just merge as-is?

Atticus1806 commented:
So to be honest I put the extension of this model on hold for now, since this is a full model which might change a number of times in the concrete implementation depending on how much returnn_common changes until its first full release. While it still has errors in the tests I think they are related to things outside of this PR. 

So we could either merge it now and update it when time comes and it is ready to use or we could leave this PR as is and update this PR then. I would be fine with both, if you want to keep the open PR's as little as possible we could merge now.

albertz commented:
I think it's easier to merge it now (or with some cleanup given my recent comments) to easier allow for changes.

Also, most things are actually ready in returnn_common to test this now. Although you are right that some things might still change.

Atticus1806 commented:
I did not really follow the development of returnn_common the last few weeks, since I was working on something else, so from my view what is open is: 

- Line 276 encoder-decoder attention. This should be `nn.dot_attention` I guess? 
- search (is this solved in general for returnn_common yet? Otherwise add this later I think)
- Handling of dimension tags (again: is this fully decided yet? How much we need to care about this here right now?)

Other things can be extended later. I think search and dimension tags might also be extended later, depending on its state right now.

albertz commented:
> * Line 276 encoder-decoder attention. This should be `nn.dot_attention` I guess?

Yes, I think so (without looking at it now).

> * search (is this solved in general for returnn_common yet? Otherwise add this later I think)

There is still some open question on the basic design but in principle it is available. But let's later look at that. Not needed for the merge.

> * Handling of dimension tags (again: is this fully decided yet? How much we need to care about this here right now?)

All on dimension tags should be clear for Transformer. They should consistently be used everywhere. There should not be `int` for dimensions. See the `Conformer` for similar code.

Atticus1806 commented:
Okay, so the commit was a bit larger than expected. Updated the dimension tags (and added defaults in the `Transformer` class)
Do we want these defaults? Or should this be mandatory to be set by the user? 
Also some naming changes.
From what I can see the only thing missing now is the "if self.search" part, but as you said this can be done later. Should we comment it out / remove it for now?

albertz commented:
Just leave it as it is. I will merge and update it.

Where do the defaults come from now? This should be documented.

Atticus1806 commented:
updated the documentation, linking the paper adding a remark for `norm_first` , since as discussed here `True` should be default.

albertz reviewed (commented):
So I cleaned up a bit. As discussed, we will just merge this now, and can the do further improvements.

albertz commented:
I think we should rename `Transformer`/`TransformerDecoder` to fit the generic high-level decoder (#49, still work-in-progress).
